### ISSUE TITLE
Button variants not accessible to screen readers

### ISSUE THREAD

**Issue by storsletten** [warp-ds/elements#728]
## Introduction
The button component does not expose the meaning conveyed by certain visual variants to screen readers. Variants that convey meaningful information beyond visual decoration include the "primary" variant and any of the "critical" variants.

For example, the "primary" variant visually emphasizes the preferred or recommended action, but this emphasis is not conveyed to screen reader users. The "critical" variants encourage sighted end-users to pay additional attention, but again this encouragement does not extend to screen readers.

A solution is to present a text description of these variants to screen readers.

## Descriptions
The following is a list of proposed external descriptions of variants that convey meaning:
- The "primary" variant should map to a localized version of the word "Highlighted".
- Any of the "critical" variants should map to a localized version of the word "Attention".

## Implementation
Map variants that convey meaning to localized descriptions and present those descriptions to screen readers using the [aria-description attribute](https://www.w3.org/TR/wai-aria-1.3/#aria-description). If the consumer of the component sets their own custom aria-description, then this should override the aria-description attribute from the component.

### Description Principles
The following is a list of guiding principles that have been applied when deciding on the external descriptions of variants that convey meaning:
- Short, preferably a single word.
- Context-agnostic, reflecting the meaning conveyed by visual styling rather than assuming intentions of consumers of the component.
- Comprehensible, easy to understand for people with reading abilities equivalent to [lower secondary education](https://www.sciencedirect.com/topics/social-sciences/lower-secondary-education).

### Rejected Description Proposals
The following is a list of external variant descriptions that have been considered and rejected:
- "Warning" or "Caution" to describe any of the "critical" variants were rejected because they may elicit apprehension rather than attention.
- "Destructive" to describe any of the "critical" variants was rejected because it assumes that all critical actions are destructive (i.e. it is not context-agnostic). It may also elicit apprehension rather than attention.
- "Prominent" to describe the "primary" variant was rejected because it is borderline for a lower secondary education audience.

### Implementation Design Considerations
The [aria-description attribute](https://www.w3.org/TR/wai-aria-1.3/#aria-description) was introduced in [ARIA 1.3](https://www.w3.org/TR/wai-aria-1.3/) and is already supported by current versions of popular screen readers including [JAWS](https://vispero.com/jaws-screen-reader-software/), [NVDA](https://www.nvaccess.org/), and [VoiceOver](https://developer.apple.com/design/human-interface-guidelines/voiceover).

Allowing consumers to override variant descriptions simplifies implementation complexity compared to a process of merging custom description and variant description, potentially across both aria-description and aria-describedby attributes, while maintaining a grammatically correct end result. Permitting the consumer to set a custom description is also better than denying custom descriptions, since custom descriptions can be assumed to be context-aware and more valuable to end users than any default description provided by the component.

There is no need to implement special handling in regards to the potential use of the [aria-describedby attribute](https://www.w3.org/TR/wai-aria-1.3/#aria-describedby) as this attribute automatically takes precedence over the aria-description attribute. This behavior is defined in the [4.2 Description Computation section of the Accessible Name and Description Computation Specification 1.2](https://www.w3.org/TR/accname-1.2/#mapping_additional_nd_description).

### Rejected Implementation Proposals
The following is a list of implementation proposals that have been considered and rejected:
- Using the [aria-roledescription attribute](https://www.w3.org/TR/wai-aria-1.3/#aria-roledescription) was rejected because it would replace existing role description instead of adding to it, resulting in worse UX. Each screen reader has its own way of presenting native roles, often in different ways depending on whether it is through speech or braille, and their users are accustomed to those specific presentation styles. For example, the screen reader JAWS will speak the button role as "button" and display it as "btn" in braille.
- Using the [aria-describedby attribute](https://www.w3.org/TR/wai-aria-1.3/#aria-describedby) was rejected because it would require a separate element solely to support a description of the button's visual variant, and there is no natural location for such an element. Placing this element inside the button could cause the description to become part of the button's accessible name, effectively duplicating the description. Placing it outside the button would expose the description independently of the button without the context that gives it meaning.
- Manipulating the button's [accessible name](https://www.w3.org/TR/accname-1.2/#dfn-accessible-name) (i.e. aria-label, aria-labelledby, or text content) was rejected because it would create considerable implementation complexity. There is currently no standard built-in JavaScript API for fetching an element's computed accessible name, and a mutation observer would also be needed in case the original accessible name is dynamically altered later after initialization. It would also be a challenge to make sure the end result is grammatically correct in all use cases after the process of merging variant description with the button's accessible name.
